src/Shared/CopyOnWriteDictionary.cs

```diff
@@ -80,54 +65,48 @@ internal CopyOnWriteDictionary(IEqualityComparer<K> keyComparer)
         /// <summary>
         /// Constructor taking a specified comparer for the keys and an initial capacity
         /// </summary>
-        internal CopyOnWriteDictionary(int capacity, IEqualityComparer<K> keyComparer)
+        internal CopyOnWriteDictionary(int capacity, IEqualityComparer<K>? keyComparer)
         {
-            this.capacity = capacity;
-            Comparer = keyComparer;
+            backing = ImmutableDictionary.Create<K, V>(keyComparer);
         }
 
         /// <summary>
         /// Serialization constructor, for crossing appdomain boundaries
         /// </summary>
-        [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "info", Justification = "Not needed")]
         [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "context", Justification = "Not needed")]
         protected CopyOnWriteDictionary(SerializationInfo info, StreamingContext context)
         {
+            object v = info.GetValue(nameof(backing), typeof(KeyValuePair<K, V>[]));
+
+            object comparer = info.GetValue(nameof(Comparer), typeof(IEqualityComparer<K>));
+
+            var b = ImmutableDictionary.Create<K, V>((IEqualityComparer<K>)comparer);
+
+            backing = b.AddRange((KeyValuePair<K, V>[])v);
         }
 
         /// <summary>
         /// Cloning constructor. Defers the actual clone.
         /// </summary>
         private CopyOnWriteDictionary(CopyOnWriteDictionary<K, V> that)
         {
-            Comparer = that.Comparer;
             backing = that.backing;
-            if (backing != null)
-            {
-                lock (((ICollection)backing).SyncRoot)
-                {
-                    backing.AddRef();
-                }
-            }
         }
 
         public CopyOnWriteDictionary(IDictionary<K, V> dictionary)
         {
-            foreach (KeyValuePair<K, V> pair in dictionary)
-            {
-                this[pair.Key] = pair.Value;
-            }
+            backing = dictionary.ToImmutableDictionary();
         }
 
         /// <summary>
         /// Returns the collection of keys in the dictionary.
         /// </summary>
-        public ICollection<K> Keys => ReadOperation.Keys;
+        public ICollection<K> Keys => ((IDictionary<K, V>)ReadOperation).Keys;
```

Comment:
Why is Keys returning an IDictionary? Same for Values. Seems like it should be a set rather than a map.

Reply:
(facepalm) It compiled because the compiler doesn't know if the object implements the interface or not unless its type is sealed. This would have failed at run-time in task host. I've fixed it by making the simple task host version of `ImmutableDictionary<K, V>` implement `IDictionary<K, V>`.

Reply:
I had to read that several times to even understand what you were saying. Very nice catch!